File: android/app/src/main/java/chat/ssh/sshchat/SecureUpload.kt
package chat.ssh.sshchat

import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.UUID

object SecureUpload {
    fun upload(url: String, key: String, file: File): String {
        val filename = file.name.replace('\\', '_').replace('/', '_').take(200).ifBlank { "file.bin" }
        val mime = MediaMime.guess(filename).let {
            if (it == "application/octet-stream") "application/octet-stream" else it
        }
        val boundary = "----SSHChat${UUID.randomUUID().toString().replace("-", "")}"
        val head = (
            "--$boundary\r\n" +
                "Content-Disposition: form-data; name=\"file\"; filename=\"$filename\"\r\n" +
                "Content-Type: $mime\r\n\r\n"
            ).toByteArray(StandardCharsets.UTF_8)
        val tail = "\r\n--$boundary--\r\n".toByteArray(StandardCharsets.UTF_8)
        val data = file.readBytes()

        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 20_000
            readTimeout = 180_000
            doInput = true
            doOutput = true
            setRequestProperty("X-Upload-Key", key.uppercase())
            setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")
            setFixedLengthStreamingMode(head.size + data.size + tail.size)
        }
        conn.outputStream.use { out ->
            out.write(head)
            out.write(data)
            out.write(tail)
        }
        val code = conn.responseCode
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        val raw = stream?.use { it.readBytes() }?.toString(StandardCharsets.UTF_8).orEmpty()
        if (code !in 200..299) {
            val err = runCatching { JSONObject(raw).optString("error") }.getOrNull()
                ?.takeIf { it.isNotBlank() }
                ?: raw.take(200).ifBlank { "HTTP $code" }
            error(err)
        }
        val payload = runCatching { JSONObject(raw.ifBlank { "{}" }) }.getOrNull()
        if (payload != null && payload.has("error") && payload.optString("error").isNotBlank()) {
            error(payload.optString("error"))
        }
        return payload?.optString("filename")?.takeIf { it.isNotBlank() } ?: filename
    }
}

File: ios/SSHChat/SecureUpload.swift
import Foundation

enum SecureUpload {
    static func upload(url: String, key: String, fileURL: URL) async throws -> String {
        let filename = fileURL.lastPathComponent
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
        let safeName = String(filename.prefix(200)).isEmpty ? "file.bin" : String(filename.prefix(200))
        var mime = MediaMime.guess(safeName)
        if mime == "application/octet-stream" { mime = "application/octet-stream" }
        let data = try Data(contentsOf: fileURL)
        let boundary = "----SSHChat\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"
        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(safeName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(mime)\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        var req = URLRequest(url: URL(string: url)!)
        req.httpMethod = "POST"
        req.timeoutInterval = 180
        req.setValue(key.uppercased(), forHTTPHeaderField: "X-Upload-Key")
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.httpBody = body

        let (respData, resp) = try await URLSession.shared.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        let raw = String(data: respData, encoding: .utf8) ?? ""
        let json = (try? JSONSerialization.jsonObject(with: respData) as? [String: Any]) ?? [:]
        if !(200...299).contains(code) {
            let err = (json["error"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            throw UploadError.message(err?.isEmpty == false ? err! : (raw.prefix(200).isEmpty ? "HTTP \(code)" : String(raw.prefix(200))))
        }
        if let err = json["error"] as? String, !err.isEmpty {
            throw UploadError.message(err)
        }
        if let remote = json["filename"] as? String, !remote.isEmpty {
            return remote
        }
        return safeName
    }

    enum UploadError: LocalizedError {
        case message(String)
        var errorDescription: String? {
            switch self {
            case .message(let s): return s
            }
        }
    }
}
